homeassistant/components/number/__init__.py

```diff
@@ -93,8 +118,56 @@ class NumberEntityDescription(EntityDescription):
 
     max_value: float | None = None
     min_value: float | None = None
+    native_max_value: float | None = None
+    native_min_value: float | None = None
+    native_unit_of_measurement: str | None = None
+    native_step: float | None = None
     step: float | None = None
 
+    def __post_init__(self) -> None:
+        """Post initialisation processing."""
+        if (
+            self.max_value is not None
+            or self.min_value is not None
+            or self.step is not None
+            or self.unit_of_measurement is not None
+        ):
+            caller = inspect.stack()[2]
+            module = inspect.getmodule(caller[0])
+            if module and module.__file__ and "custom_components" in module.__file__:
+                report_issue = "report it to the custom component author."
+            else:
+                report_issue = (
+                    "create a bug report at "
+                    "https://github.com/home-assistant/core/issues?q=is%3Aopen+is%3Aissue"
+                )
+            _LOGGER.warning(
+                "%s is setting deprecated attributes on an instance of "
+                "NumberEntityDescription, this is not valid and will be unsupported "
+                "from Home Assistant 2022.10. Please %s",
+                module.__name__ if module else self.__class__.__name__,
+                report_issue,
+            )
+            self.native_unit_of_measurement = self.unit_of_measurement
+
+
+def ceil_decimal(value: float, precision: float = 0) -> float:
```

Comment:
do we have similar logic for sensors? should we make a helper for these things?

Reply:
We don't need to take the ceiling or floor for sensors.
For sensors we instead have a funny way of rounding where we round to no decimals, i.e. integers

I think what we should do is differentiate between the precision and the friendly string representation, but that's for another PR.